azkaban-common/src/main/java/azkaban/executor/container/KubernetesContainerizedImpl.java

```diff
@@ -1131,6 +1141,54 @@ private void createPod(final int executionId)
     if (flowParam != null && !flowParam.isEmpty()) {
       logger.info("ExecId: {}, Flow Parameters are: {}", executionId, flowParam);
     }
+
+    /*
+     The Set proxyUsersMap is being created with the intention of determining all the proxy users
+     prior to pod dispatch.
+     Why this is needed ?
+     -> When we use POLP(Principle of Least Priviledge) , initially introduced in
+     https://github.com/azkaban/azkaban/pull/3216, we need to have a list of users that custom
+     credentials (getCustomCredentialProvider) need to be fetched and registered for, as we will
+     not be able to fetch them at runtime as Azkaban user present in the Azkaban Executor Pod
+     will no longer be able to do so. This prefetching of custom credentials  is performed in
+     the azkaban-security-init container by attaching the proxy users needed to be prefetched as
+     environment variables.
+     */
+    Set<String> proxyUsersMap = new HashSet<>();
+
+    final boolean prefetchAllProxyUserCredentials =
+        flowParam != null && flowParam.containsKey(FlowParameters.PROXY_USER_PREFETCH_ALL) &&
+            Boolean.parseBoolean(flowParam.get(FlowParameters.PROXY_USER_PREFETCH_ALL));
+    /*
+     As an optimization for not walking through the DAG and making DB queries for large DAGs we
+     try to check if the proxy users list from the project page is less then a certain threshold.
+     If it is, less than this threshold, we simply add all of them. This configuration will be
+     custom to a given environment. An option to fetch all of them via the project page is
+     also provided with a flow parameter.
+     */
+
+    if (flow.getProxyUsers().size() > this.proxyUserPrefetchThreshold || !prefetchAllProxyUserCredentials) {
+        proxyUsersMap.addAll(ContainerImplUtils
+            .getProxyUsersForFlow(this.projectManager, flow));
+      }
+      else{
+        proxyUsersMap.addAll(flow.getProxyUsers());
+      }
+    // If certain jobtypes need specific user credentials we add them to the prefetch list
+    if (this.jobTypePrefetchUserMap != null) {
+      final Set<String> jobTypeUsersForFlow =
+          ContainerImplUtils.getJobTypeUsersForFlow(this.jobTypePrefetchUserMap, jobTypes);
+          proxyUsersMap.addAll(jobTypeUsersForFlow);
+      }
```

Comment:
It can be addressed in a separate PR but perhaps the list of jobtype proxy users can fetched from the pod spec template instead of adding them to the web server configuration. The web server probably shouldn't be aware of jobtype details and by leveraging the pod spec template the web server config is more clean. 

The pod spec template can also be leveraged to specify proxy users for which tokens always need to be prefetched (ie: the user to download the project zips). Where are those required proxy users defined right now?

Reply:
_The pod spec template can also be leveraged to specify proxy users for which tokens always need to be prefetched_

That is the plan ^ . 
However for the jobtype proxy users we need to programmatically determine if that jobtype is present. it cannot be present by default in the pod spec template.

Reply:
Great!
Maybe I didn't explain it well, I'm just suggesting to move the web server config `prefetch.jobtype.user.map` to the pod spec template. In the code, we would fetch that info from the template instead of the webserver config here:
```
this.jobTypePrefetchUserMap =
      this.azkProps.getString(ContainerizedDispatchManagerProperties.PREFETCH_JOBTYPE_PROXY_USER_MAP, null);
```
Everything else remains the same.

Reply:
Of course the jobtype proxy users won't go in the same place as the always-present proxy users.